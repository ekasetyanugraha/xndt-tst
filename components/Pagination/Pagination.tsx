import UrlParamsUniversity from "interfaces/url-params-university";

const DEFAULT_PAGES = 5;

interface Props {
  limit?: number;
  offset?: number;
  total?: number;
  onChangePage?: (page: UrlParamsUniversity) => void;
}

export default function Pagination({
  limit = 0,
  offset = 0,
  total = 0,
  onChangePage = () => {},
}: Props) {
  const currentPage = (offset / limit) + 1;
  const totalPage = total && limit ? Math.ceil(total / limit) : 0;
  const totalPageLimited = totalPage < DEFAULT_PAGES ? totalPage : DEFAULT_PAGES;

  const disablePrev = currentPage === 1;
  const disableNext = currentPage === totalPage;

  const changePage = (page: number) => onChangePage({
    limit,
    offset: (page - 1) * limit,
  });

  const onClickPrev = () => {
    changePage(currentPage - 1);
  };

  const onClickNext = () => {
    changePage(currentPage + 1);
  };

  return (
    <nav className="pagination is-centered" role="navigation" aria-label="pagination">
      <button
        type="button"
        className="button pagination-previous"
        title="This is the first page"
        disabled={disablePrev}
        onClick={onClickPrev}
      >
        Previous
      </button>

      <ul className="pagination-list">
        {
          [...Array(totalPageLimited)].map((page, i) =>
            <li key={i}>
              <button
                type="button"
                className={`button pagination-link ${currentPage === i + 1 && 'is-current'}`}
                onClick={() => changePage(i + 1)}
              >
                {i + 1}
              </button>
            </li>
          )
        }
      </ul>

      <button
        type="button"
        className="button pagination-next"
        disabled={disableNext}
        onClick={onClickNext}
      >
        Next page
      </button>
    </nav>
  )
}
