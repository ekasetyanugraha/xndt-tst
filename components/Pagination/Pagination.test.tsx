import React from 'react';
import { render, screen } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import Pagination from '.';

describe('Component: Pagination', () => {
  describe('without props', () => {
    beforeEach(() => {
      render(<Pagination />);
    });

    it('should match snapshot', () => {
      expect(screen.getByLabelText('pagination')).toMatchSnapshot();
    });

    it('should have previous button', () => {
      expect(screen.getByTestId('button-previous')).toBeTruthy();
    });

    it('should have next button', () => {
      expect(screen.getByTestId('button-next')).toBeTruthy();
    });

    it('should not have any page number button', () => {
      expect(screen.queryByTestId('button-page')).toBeNull();
    });
  });

  describe('more than 5 pages', () => {
    const mockLimit = 10;
    const mockOffset = 30;
    const mockTotal = 200;
    const mockOnChangePage = jest.fn();

    beforeEach(() => {
      render(<Pagination limit={mockLimit} offset={mockOffset} total={mockTotal} onChangePage={mockOnChangePage} />);
    });

    it('should have previous button', () => {
      expect(screen.getByTestId('button-previous')).toBeTruthy();
    });

    it('should have next button', () => {
      expect(screen.getByTestId('button-next')).toBeTruthy();
    });

    it(`should 5 page number button`, () => {
      expect(screen.getAllByTestId('button-page')).toHaveLength(5);
    });

    describe('when user clicks previous button', () => {
      beforeEach(() => {
        const prevButton = screen.getByTestId('button-previous');
        userEvent.click(prevButton);
      });

      it('should call onChangePage with the right payload', () => {
        expect(mockOnChangePage).toHaveBeenLastCalledWith({
          limit: mockLimit,
          offset: mockOffset - mockLimit,
        });
      });
    });

    describe('when user clicks next button', () => {
      beforeEach(() => {
        const nextButton = screen.getByTestId('button-next');
        userEvent.click(nextButton);
      });

      it('should call onChangePage with the right payload', () => {
        expect(mockOnChangePage).toHaveBeenLastCalledWith({
          limit: mockLimit,
          offset: mockOffset + mockLimit,
        });
      });
    });

    describe('when user clicks page number button', () => {
      const mockPage = 2;

      beforeEach(() => {
        const pageButton = screen.getAllByTestId('button-page')[mockPage - 1];
        userEvent.click(pageButton);
      });

      it('should call onChangePage with the right payload', () => {
        expect(mockOnChangePage).toHaveBeenLastCalledWith({
          limit: mockLimit,
          offset: (mockPage - 1) * mockLimit,
        });
      });
    });
  });

  describe('less than 5 pages', () => {
    const mockLimit = 10;
    const mockOffset = 30;
    const mockTotal = 40;

    beforeEach(() => {
      render(<Pagination limit={mockLimit} offset={mockOffset} total={mockTotal} />);
    });

    it('should have previous button', () => {
      expect(screen.getByTestId('button-previous')).toBeTruthy();
    });

    it('should have next button', () => {
      expect(screen.getByTestId('button-next')).toBeTruthy();
    });

    it(`should 4 page number button`, () => {
      expect(screen.getAllByTestId('button-page')).toHaveLength(4);
    });
  });
});
