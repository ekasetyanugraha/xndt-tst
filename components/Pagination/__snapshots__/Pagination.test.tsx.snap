// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Component: Pagination without props should match snapshot 1`] = `
<nav
  aria-label="pagination"
  class="pagination is-centered"
  role="navigation"
>
  <button
    class="button pagination-previous"
    data-testid="button-previous"
    title="This is the first page"
    type="button"
  >
    Previous
  </button>
  <ul
    class="pagination-list"
  />
  <button
    class="button pagination-next"
    data-testid="button-next"
    type="button"
  >
    Next page
  </button>
</nav>
`;
